File: internal/adapter/adapter.go
// Package adapter turns one task execution's captured output into a result
// envelope, by running the adapter the task named.
//
// An adapter is a separate process and it reaches the verdict. Bolt decides
// whether an execution passed in exactly three cases, each named where it
// arises: an execution bolt terminated, an adapter that reached no
// authoritative result, and a task that did not produce the evidence it
// declared.
package adapter

import (
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
)

// Envelope names the file an adapter writes. The path is the work directory it
// was given and the name never varies, so no flag says where the envelope goes
// and no task can put it somewhere else.
const Envelope = "output.yaml"

// ExitCode is the generic adapter every task gets when it names none. Its name
// is not a path: it is resolved before the config directory is consulted, so a
// project cannot shadow it by accident.
const ExitCode = "exit-code"

// A Result says what happened to the adapter itself, not what it concluded.
// What it concluded is in the envelope it wrote.
type Result struct {
	// Wrote is true when an envelope is on disk to be read.
	Wrote bool
	// Reason, when Wrote is false, says which of FR-6.11's three happened.
	Reason string
	// Kind pairs with Reason so a consumer tells them apart without reading
	// English.
	Kind string
}

// Locations are the three directories an adapter is handed, the same three
// every task gets.
type Locations struct {
	ProjectRoot string
	BaseDir     string
	WorkDir     string
}

// Resolve finds the executable for an adapter named by a task.
//
// It comes from the config directory, where jigs already come from, so a jig
// and the adapters it names travel together and `link-jigs` places both or
// neither.
func Resolve(configDir, name string) (string, error) {
	if name == "" || name == ExitCode {
		return "", nil
	}

	path := filepath.Join(configDir, name)
	info, err := os.Stat(path)
	if err != nil {
		return "", fmt.Errorf("adapter %q is not in %s: %w", name, configDir, err)
	}
	if info.IsDir() {
		return "", fmt.Errorf("adapter %q in %s is a directory", name, configDir)
	}
	if info.Mode()&0o111 == 0 {
		return "", fmt.Errorf("adapter %q in %s is not executable", name, configDir)
	}
	return path, nil
}

// DefaultArgs is the invocation bolt writes when a task does not write its own.
//
// It names the captured files and the three locations. Evidence is repeated
// once per declared file, because a task declares its evidence and those are
// what --evidence names; discovery would hand an adapter whatever a tool
// happened to leave behind.
func DefaultArgs(locations Locations, evidence []string) []string {
	args := []string{
		"--stdout", filepath.Join(locations.WorkDir, "stdout"),
		"--stderr", filepath.Join(locations.WorkDir, "stderr"),
		"--exitcode", filepath.Join(locations.WorkDir, "exitcode"),
		"--project-root", locations.ProjectRoot,
		"--base-dir", locations.BaseDir,
		"--work-dir", locations.WorkDir,
	}
	for _, path := range evidence {
		args = append(args, "--evidence", filepath.Join(locations.WorkDir, path))
	}
	return args
}

// MissingEvidence returns the declared evidence files the task did not
// produce, relative to the work directory.
//
// A task declaring evidence it did not write did not do what it said, and the
// refusal to discover means nothing else notices.
func MissingEvidence(workDir string, evidence []string) []string {
	var missing []string
	for _, path := range evidence {
		if _, err := os.Stat(filepath.Join(workDir, path)); err != nil {
			missing = append(missing, path)
		}
	}
	return missing
}

// Run invokes the adapter with the invocation bolt wrote.
func Run(executable string, args []string, locations Locations) (Result, error) {
	return finish(exec.Command(executable, args...), locations)
}

// RunShell invokes an explicit adapter invocation, written by the task in place
// of the default one. It is a shell line like a command, and the envelope is
// still expected at the same place, because no flag says where it goes.
func RunShell(command string, locations Locations) (Result, error) {
	return finish(exec.Command("sh", "-c", command), locations)
}

// finish runs a prepared adapter process and says whether an authoritative
// envelope resulted.
//
// It removes any envelope already in the work directory first, so "the adapter
// wrote one" cannot be satisfied by a leftover from a previous fold.
func finish(process *exec.Cmd, locations Locations) (Result, error) {
	envelope := filepath.Join(locations.WorkDir, Envelope)
	if err := os.Remove(envelope); err != nil && !errors.Is(err, os.ErrNotExist) {
		return Result{}, err
	}

	process.Dir = locations.BaseDir
	process.Stdin = nil

	// An adapter's own streams are captured beside the command's, so a broken
	// adapter can be read rather than guessed at.
	output, runErr := process.CombinedOutput()
	if writeErr := os.WriteFile(
		filepath.Join(locations.WorkDir, "adapter-output"), output, 0o644,
	); writeErr != nil {
		return Result{}, writeErr
	}

	if runErr != nil {
		var exit *exec.ExitError
		if !errors.As(runErr, &exit) {
			return Result{}, fmt.Errorf("running adapter %s: %w", process.Path, runErr)
		}
		return Result{
			Kind:   "adapter-failed",
			Reason: fmt.Sprintf("the adapter exited %d; its output is in adapter-output", exit.ExitCode()),
		}, nil
	}

	if _, err := os.Stat(envelope); err != nil {
		return Result{
			Kind:   "adapter-wrote-nothing",
			Reason: fmt.Sprintf("the adapter exited 0 and wrote no %s", Envelope),
		}, nil
	}

	return Result{Wrote: true}, nil
}

File: src/adapter.rs
//! Turning one execution's captured output into a verdict.
//!
//! FR-6.1: an adapter is a separate process, and where it reached an
//! authoritative result that result **is** the verdict. Bolt does not
//! second-guess one.
//!
//! FR-6.1a says when bolt writes an envelope itself, and says it as a rule
//! rather than a count: only where no adapter's result is available to take.
//! FR-6.1b records why it stopped counting, which is that counting was wrong
//! twice and a list claiming completeness invites the next reader to trust the
//! number rather than the rule.

use std::path::{Path, PathBuf};

/// The name of the file holding an execution's captured standard output.
pub const STDOUT_FILE: &str = "stdout";

/// The name of the file holding an execution's captured standard error.
pub const STDERR_FILE: &str = "stderr";

/// Why bolt wrote an envelope instead of taking an adapter's, by FR-6.11.
///
/// The three are kept apart because they have different causes: a crashing
/// adapter, a silent one, and one whose output is not an envelope are three
/// different things to go and fix.
#[derive(Debug, Clone, Copy)]
pub enum Unauthoritative {
    /// The adapter ran and exited non-zero.
    Exited(i32),
    /// The adapter left no `output.yaml` where FR-6.2b says it goes.
    WroteNothing,
    /// It left one that will not parse or will not validate.
    WroteInvalid,
}

impl Unauthoritative {
    /// The reason `kind` for this case, by FR-7.9.
    ///
    /// A consumer tells one from another without reading English, which is what
    /// FR-7.10 rests on.
    #[must_use]
    pub fn kind(self) -> &'static str {
        match self {
            Self::Exited(_) => "adapter-failed",
            Self::WroteNothing => "adapter-wrote-nothing",
            Self::WroteInvalid => "adapter-wrote-invalid",
        }
    }

    /// What went wrong, in the words a person reads.
    #[must_use]
    pub fn message(self, adapter: &str) -> String {
        match self {
            Self::Exited(status) => format!("the adapter {adapter} exited {status}"),
            Self::WroteNothing => {
                format!("the adapter {adapter} wrote no {}", crate::run::OUTPUT_FILE)
            }
            Self::WroteInvalid => format!(
                "the adapter {adapter} wrote a {} that is not an envelope",
                crate::run::OUTPUT_FILE
            ),
        }
    }
}

/// Where an adapter named `name` is found, by FR-6.10.
///
/// Resolved from the config directory, where FR-2.8 already finds jigs, so a jig
/// and the adapters it names travel together and `link-jigs` places both or
/// neither.
#[must_use]
pub fn path(config_dir: &Path, name: &str) -> PathBuf {
    config_dir.join(name)
}

/// FR-6.2's default invocation, as a shell line.
///
/// Names the captured files and the locations an adapter is handed. FR-6.2a
/// gives it the same three locations every task gets; FR-6.2c has `--evidence`
/// name what the task declared and nothing it did not.
///
/// FR-6.3 hands the exit code over as a file rather than as a verdict: whether
/// that number explains anything is the adapter's judgement, not bolt's.
///
/// No flag says where the envelope goes, by FR-6.2b. The path is the work
/// directory the adapter was given and the name never varies.
#[must_use]
pub fn default_invocation(adapter: &str, evidence: &[String]) -> String {
    let mut line = format!(
        "{} --stdout {{work_dir}}/{STDOUT_FILE} --stderr {{work_dir}}/{STDERR_FILE} \
         --exitcode {{work_dir}}/{} --project-root {{project_root}} --base-dir {{base_dir}} \
         --work-dir {{work_dir}}",
        shell_word(adapter),
        crate::run::EXITCODE_FILE,
    );
    // FR-6.2c: one `--evidence` per declared file, so an adapter reading two
    // artifacts is told about both and about nothing else.
    for file in evidence {
        line.push_str(" --evidence {work_dir}/");
        line.push_str(&shell_word(file));
    }
    line
}

/// A word safe to put on a shell line without changing what it means.
///
/// The adapter name and the evidence filenames come from a jig, which is the
/// project's own file, but they still reach a shell. FR-4.3's reasoning applies
/// wherever bolt builds a command line rather than only where a *path* is
/// substituted, and the substitution pass this line goes through afterwards is
/// single-pass by design, so a quoted word stays one word.
fn shell_word(value: &str) -> String {
    if value
        .chars()
        .all(|c| c.is_ascii_alphanumeric() || "._-/".contains(c))
    {
        return value.to_owned();
    }
    crate::selection::quote_str(value)
}
